zenoh-ext/src/advanced_subscriber.rs

```diff
@@ -0,0 +1,1316 @@
+    }
+
+    /// Enable query for historical data.
+    ///
+    /// History can only be retransmitted by Publishers that enable caching.
+    #[zenoh_macros::unstable]
+    #[inline]
+    pub fn history(mut self, config: HistoryConfig) -> Self {
+        self.history = Some(config);
+        self
+    }
+
+    /// Allow this subscriber to be detected through liveliness.
+    #[zenoh_macros::unstable]
+    pub fn subscriber_detection(mut self) -> Self {
+        self.liveliness = true;
+        self
+    }
+
+    /// A key expression added to the liveliness token key expression.
+    /// It can be used to convey meta data.
+    #[zenoh_macros::unstable]
+    pub fn subscriber_detection_metadata<TryIntoKeyExpr>(mut self, meta: TryIntoKeyExpr) -> Self
+    where
+        TryIntoKeyExpr: TryInto<KeyExpr<'c>>,
+        <TryIntoKeyExpr as TryInto<KeyExpr<'c>>>::Error: Into<zenoh::Error>,
+    {
+        self.meta_key_expr = Some(meta.try_into().map_err(Into::into));
+        self
+    }
+
+    #[zenoh_macros::unstable]
+    fn with_static_keys(self) -> AdvancedSubscriberBuilder<'a, 'static, 'static, Handler> {
+        AdvancedSubscriberBuilder {
+            session: self.session,
+            key_expr: self.key_expr.map(|s| s.into_owned()),
+            origin: self.origin,
+            retransmission: self.retransmission,
+            query_target: self.query_target,
+            query_timeout: self.query_timeout,
+            history: self.history,
+            liveliness: self.liveliness,
+            meta_key_expr: self.meta_key_expr.map(|s| s.map(|s| s.into_owned())),
+            handler: self.handler,
+        }
+    }
+}
+
+#[zenoh_macros::unstable]
+impl<Handler> Resolvable for AdvancedSubscriberBuilder<'_, '_, '_, Handler>
+where
+    Handler: IntoHandler<Sample>,
+    Handler::Handler: Send,
+{
+    type To = ZResult<AdvancedSubscriber<Handler::Handler>>;
+}
+
+#[zenoh_macros::unstable]
+impl<Handler> Wait for AdvancedSubscriberBuilder<'_, '_, '_, Handler>
+where
+    Handler: IntoHandler<Sample> + Send,
+    Handler::Handler: Send,
+{
+    #[zenoh_macros::unstable]
+    fn wait(self) -> <Self as Resolvable>::To {
+        AdvancedSubscriber::new(self.with_static_keys())
+    }
+}
+
+#[zenoh_macros::unstable]
+impl<Handler> IntoFuture for AdvancedSubscriberBuilder<'_, '_, '_, Handler>
+where
+    Handler: IntoHandler<Sample> + Send,
+    Handler::Handler: Send,
+{
+    type Output = <Self as Resolvable>::To;
+    type IntoFuture = Ready<<Self as Resolvable>::To>;
+
+    #[zenoh_macros::unstable]
+    fn into_future(self) -> Self::IntoFuture {
+        std::future::ready(self.wait())
+    }
+}
+
+#[zenoh_macros::unstable]
+struct Period {
+    timer: Timer,
+    period: Duration,
+}
+
+#[zenoh_macros::unstable]
+struct State {
+    next_id: usize,
+    global_pending_queries: u64,
+    sequenced_states: HashMap<EntityGlobalId, SourceState<u32>>,
+    timestamped_states: HashMap<ID, SourceState<Timestamp>>,
+    session: Session,
```

Comment:
State should not have a `Session` but a `WeakSession`, otherwise, keeping a reference to an advanced subscriber prevent the session to be closed, contrary to a normal subscriber.